### PR TITLE
Conversion of calldata array slices to memory

### LINKED ISSUE #8991: [codegen] ICE because _targetType != typeOnStack.arrayType();
## Description

```
contract C {
  function f(int[] calldata b) external { [b[:]]; }
}
```

throws

https://github.com/ethereum/solidity/blob/678a801daf44b563272ed9c699caac8087dc7ee1/libsolidity/codegen/CompilerUtils.cpp#L1017

## Environment

- Compiler version: latest develop
## Steps to Reproduce

```
$ solc <input.sol>
```

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #8991 for both old codegen and IR.

Test coverage for array slices is not great, so I guess will need to add more tests here

### CODE REVIEW

**Review by ekpyron — COMMENTED** [review]
This really also implements conversion from calldata array slices to storage, doesn't it? Can you add a test for that?

File: test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_storage.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+contract C {
+    int[][] s;
+    function f(int[] calldata b, uint256 start, uint256 end) public returns (int) {
+        s.push();
+        s[0] = int[](b[start:end]);
```
Comment (mijovic) [inline test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_storage.sol]: @ekpyron 

Without explicitly converting to `int[]` this would not pass `TypeChecker`. That makes this first converting to calldata array, than copies to storage, which is done trough already implemented code of array conversion.

Does it makes sense to allow implicit conversion of slices to arrays as I think it will work just fine? (Of course, this should be done as a breaking change)

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -2074,7 +2074,13 @@ std::unique_ptr<ReferenceType> ArrayType::copyForLocation(DataLocation _location
 
 BoolResult ArraySliceType::isImplicitlyConvertibleTo(Type const& _other) const
 {
-	if (m_arrayType.location() == DataLocation::CallData && m_arrayType.isDynamicallySized() && m_arrayType == _other)
+	if (
+		auto const* otherArrayType = dynamic_cast<ArrayType const*>(&_other);
+		otherArrayType &&
+		m_arrayType.location() == DataLocation::CallData &&
+		m_arrayType.isDynamicallySized() &&
+		*otherArrayType == *m_arrayType.copyForLocation(otherArrayType->location(), false)
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Can this just be ``m_arrayType->isImplicitlyConvertibleTo(_other)``?
I mean, we basically implement it by calling the array conversion functions, so all cases in which that is fine, should be fine here, shouldn't they?

Comment (mijovic) [reply libsolidity/ast/Types.cpp]: It should work for all cases implemented in array conversion function, yes. Also, coverage there will be full after closing currently open PRs.
I updated to what you suggested, looks much better :-)

File: test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_storage.sol (hunk 1 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_storage.sol]: We discussed in channel, that this PR should allow conversions as we don't see any problem with it.

If we end up with not allowing this implicit conversion, we agreed that, also the weird inline case should be disallowed as well.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 3]
```diff
@@ -1051,11 +1051,11 @@ void CompilerUtils::convertType(
...
-			_targetType.dataStoredIn(DataLocation::CallData),
-			"Conversion from calldata slices to memory not yet implemented."
-		);
-		solAssert(_targetType == typeOnStack.arrayType(), "");
+		solAssert(_targetType.category() == Type::Category::Array, "");
+		auto const& targetArrayType = dynamic_cast<ArrayType const&>(_targetType);
+		solAssert(targetArrayType == *typeOnStack.arrayType().copyForLocation(targetArrayType.location(), false), "");
+		if (!_targetType.dataStoredIn(DataLocation::CallData))
+			return convertType(typeOnStack.arrayType(), _targetType);
 		solUnimplementedAssert(
```
Comment (ekpyron) [inline libsolidity/codegen/CompilerUtils.cpp]: This assertion is generally relevant and will be true for all valid cases, won't it? I think it should move up before the early return, just to be sure.

**Review by ekpyron — COMMENTED** [review]
Looks good to me now apart from the one remaining comment.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -2074,7 +2074,11 @@ std::unique_ptr<ReferenceType> ArrayType::copyForLocation(DataLocation _location
...
 BoolResult ArraySliceType::isImplicitlyConvertibleTo(Type const& _other) const
 {
-	if (m_arrayType.location() == DataLocation::CallData && m_arrayType.isDynamicallySized() && m_arrayType == _other)
+	if (
+		m_arrayType.dataStoredIn(DataLocation::CallData) &&
+		m_arrayType.isDynamicallySized() &&
+		m_arrayType.isImplicitlyConvertibleTo(_other)
+	)
 		return true;
 	return (*this) == _other;
```
Comment (cameel) [inline libsolidity/ast/Types.cpp]: Since these are all simple boolean checks, how about this style?

```suggestion
	return
		(*this) == _other ||
		m_arrayType.dataStoredIn(DataLocation::CallData) &&
		m_arrayType.isDynamicallySized() &&
		m_arrayType.isImplicitlyConvertibleTo(_other);
```

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 5]
```diff
@@ -2852,15 +2852,18 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
...
 			*fromType.arrayType().baseType() == *targetType.baseType(),
 			"Converting arrays of different type is not possible"
 		);
 
+		if (!targetType.dataStoredIn(DataLocation::CallData))
+			return arrayConversionFunction(fromType.arrayType(), targetType);
+
 		string const functionName =
 			"convert_" +
 			_from.identifier() +
```
Comment (cameel) [inline libsolidity/codegen/YulUtilFunctions.cpp:2868]: I see that the code below, in the `else if (_from.category() == Type::Category::Array)` branch, does not have any special case for calldata -> calldata case. Shouldn't it be handled just like the slice here? If so, maybe it would make sense to move the code below that handles calldata -> calldata to `arrayConversionFunction()` and handle them both there?

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol [hunk 6]
```diff
@@ -4,4 +4,3 @@ contract C {
     }
 }
 // ----
-// TypeError 9553: (79-85): Invalid type for argument in function call. Invalid implicit conversion from bytes calldata slice to bytes memory requested.
```
Comment (cameel) [inline test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol:7]: I'm not sure I fully understand what is happening there. The error message implies that a calldata -> memory conversion is happening in `this.f(x[1:2])`. The function accepts `bytes calldata` so there must be also a memory -> calldata conversion after that. How does that even work? Isn't the memory -> calldata part impossible since calldata is immutable?

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol [hunk 7]
```diff
@@ -1,7 +1,5 @@
 contract C {
-    function f(bytes calldata x) external {
-        bytes memory y = x[1:2];
+    function f(bytes calldata x) external pure returns (bytes memory) {
+        return x[1:2];
```
Comment (cameel) [inline test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol:3]: Why was this change necessary? Is it purely stylistic or does the difference actually matter? If it does, then we could keep the `bytes memory y = x[1:2];` line inside the function to test both.

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol [hunk 8]
```diff
@@ -1,7 +1,5 @@
 contract C {
-    function f(bytes calldata x) external {
-        bytes memory y = x[1:2];
```
Comment (cameel) [inline test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol:3]: Do we have a test for calldata slice -> calldata array conversion? If not, please add one.

File: test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_memory.sol [hunk 9]
```diff
@@ -0,0 +1,29 @@
+contract C {
```
Comment (cameel) [inline test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_memory.sol:1]: Could you also add `array_slice_calldata_to_calldata.sol` to make sure that the `calldata` -> `calldata` case works too?

Also, given the assertions in the old codegen, maybe it would be a good idea to have a test for a case with more complex nesting. E.g. slicing a dynamic array of structs containing static arrays?

File: libsolidity/codegen/CompilerUtils.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/codegen/CompilerUtils.cpp]: Yeah. Though maybe it should just be an ordinary assert because this seems to be covered by the type checker. `solUnimplementedAssert()` suggests that this code would actually be reachable.

On 0.7.5 slicing `uint[][] calldata` results in:
```
Error: Index range access is not supported for arrays with dynamically encoded base types.
```

Slicing `uint[][10] calldata` or `uint[10][10] calldata` results in:
```
Error: Index range access is only supported for dynamic calldata arrays.
```

File: libsolidity/codegen/CompilerUtils.cpp [hunk 10]
```diff
@@ -1051,11 +1051,11 @@ void CompilerUtils::convertType(
...
 	{
 		auto& typeOnStack = dynamic_cast<ArraySliceType const&>(_typeOnStack);
-		solUnimplementedAssert(
-			_targetType.dataStoredIn(DataLocation::CallData),
-			"Conversion from calldata slices to memory not yet implemented."
-		);
-		solAssert(_targetType == typeOnStack.arrayType(), "");
+		solAssert(_targetType.category() == Type::Category::Array, "");
+		auto const& targetArrayType = dynamic_cast<ArrayType const&>(_targetType);
+		solAssert(targetArrayType == *typeOnStack.arrayType().copyForLocation(targetArrayType.location(), false), "");
```
Comment (cameel) [inline libsolidity/codegen/CompilerUtils.cpp]: I think this condition is more restrictive than the one in `ArraySliceType::isImplicitlyConvertibleTo()` so there will be some cases that pass through type checker just fine and fail here.

Looking at [`ArrayType::isImplicitlyConvertibleTo()`](https://github.com/ethereum/solidity/blob/v0.7.5/libsolidity/ast/Types.cpp#L1684-L1722) I think that for example `int[] calldata` -> `uint[] storage` should be one such types.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 3 above)
Comment (ekpyron) [reply libsolidity/codegen/CompilerUtils.cpp]: Yes, it indeed should be a regular assert - unfortunately at some point we realized that slicing for multi-dimensional arrays unfortunately will never be possible with the current way ABI encoding works.

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol (hunk 6 above)
Comment (ekpyron) [reply test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol:7]: That's how all external calls work - arguments to ``CALL`` are in memory and on the other end of the call arrive in calldata.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 5 above)
Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.cpp:2868]: Are there any calldata -> calldata conversions at all?

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol (hunk 6 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol:7]: Thanks! Yeah, seems kinda obvious when I look at from the EVM perspective. No way to address calldata in `CALL`.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 5 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp:2868]: I'm not sure, I originally thought that [calldata_dynamic_forward.sol](https://github.com/ethereum/solidity/blob/develop/test/libsolidity/syntaxTests/array/slice/calldata_dynamic_forward.sol) should be an example of that but, as you explaned, that's an external call and goes through a `calldata` -> `memory` conversion.

In that case how about passing calldata slice to an internal function that accepts a calldata array?

If that's not possible then this is just a piece of dead code and should actually be removed.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/codegen/CompilerUtils.cpp]: Multi-dimensional dynamic arrays or arrays in general?

Comment (ekpyron) [reply libsolidity/codegen/CompilerUtils.cpp]: Yes - arrays of dynamic array address the inner array using a tail pointer that's relative to the "base reference", i.e. relative to the entire array. A slice doesn't have that reference anymore, unless we'd store it somewhere, which would make them three-stack-slot-types which would be rather silly... one can definitely work around it, but unclear, if the cost/benefit ratio of that both implementation-effort- and gas-wise will be justified in the end...

Comment (ekpyron) [reply libsolidity/codegen/CompilerUtils.cpp]: dynamic or static arrays of static arrays could still work though - that part one might just classify as "unimplemented"...

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 5 above)
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp:2868]: conversion to calldata doesn't happen, except `ArraySliceType` -> `ArrayType` which happens here. That is happening when doing index access for calldata array slice, for example in https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/array/calldata_slice_access.sol

I would not add `calldata` to `calldata` case in `arrayConversionFunction` as it can be misleading. Regarding test what @cameel suggested, I added it and there is no conversion from `calldata` -> `calldata` at all. Only in case from array slice -> calldata array. Also, conversions in calldata are not possible. For example passing `uint128[]` as an argument to function accepting `uint256[]` is not possible (`TypeChecker` doesn't allow that)

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol (hunk 7 above)
Comment (mijovic) [reply test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol:3]: I added it to get rid of unused local variable :)

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol (hunk 8 above)
Comment (mijovic) [reply test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol:3]: In here we have that. 
https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/array/calldata_slice_access.sol

But, I am adding new test as well

File: libsolidity/codegen/CompilerUtils.cpp (hunk 10 above)
Comment (mijovic) [reply libsolidity/codegen/CompilerUtils.cpp]: I have changed assertions in both places to be the same, and to align to `TypeChecker`. Will add more tests now, for converting of base types that are not the same.

**Comment by mijovic** [comment]
@cameel I think I have cleaned up most of the things and added few more tests

File: test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_memory.sol (hunk 9 above)
Comment (mijovic) [reply test/libsolidity/semanticTests/array/slices/array_slice_calldata_to_memory.sol:1]: Added such test

**Review by chriseth — CHANGES_REQUESTED** [review]
Would be nice to have some more tests, like conversion as part of external function calls, as part of `abi.encode`, `abi.decode`, `address.call` and `keccak256`.

File: test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol (hunk 7 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/array/slice/calldata_dynamic_convert_to_memory.sol:3]: Oh It's fine then. :)

**Review by cameel — DISMISSED** [review]
All my concerns have been addressed so I'm marking as approved. I think it's fine to merge once you add the tests requested by @chriseth.

**Comment by mijovic** [comment]
Added test for calling `keccak256` with array slice as argument. Will add other calls as well in the morning.

**Comment by mijovic** [comment]
Added tests for `abi.decode`, `abi.encode` and `keccak256`

**Review by chriseth — APPROVED** [review]
